src/main/java/net/objecthunter/exp4j/tokenizer/Tokenizer.java

```diff
@@ -133,6 +148,16 @@ private boolean isCloseParentheses(char ch) {
         return ch == ')' || ch == '}' || ch == ']';
     }
 
+    private Collection<String> undefinedVariables = null;
+    
+    public void allowUndefinedVariables(boolean allow) {
+    	this.undefinedVariables = (allow ? new ArrayList<String>() : null);
```

Comment:
I think you could use ```HashSet```, or rather ```LinkedHashSet``` if you want to preserve tokens order. ```ArrayList``` will consume less memory but ```HashSet``` will perform better. I'm not sure what is more important here.

The ```ArrayList```:
* contains - O(N)
* add - O(1)

The ```HashSet``` or ```LinkedHashSet```:
* contains - O(1)
* add - O(1)

But also, with ```HashSet``` you could not check ```contains(name)``` since ```Set``` guarantees ```String```s uniqueness on ```add(name)``` so the number of operations is smaller.